src/main/core/logger/shadow_logger.rs

```diff
@@ -0,0 +1,400 @@
+use once_cell::sync::Lazy;
+use std::cell::RefCell;
+use std::convert::TryFrom;
+use std::sync::mpsc::{Receiver, Sender};
+use std::sync::{Mutex, RwLock};
+use std::time::Duration;
+
+/// Trigger an asynchronous flush when this many lines are queued.
+const ASYNC_FLUSH_QD_LINES_THRESHOLD: usize = 100_000;
+
+/// Performs a *synchronous* flush when this many lines are queued.  i.e. if
+/// after reaching the `ASYNC_FLUSH_QD_LINES_THRESHOLD`, log lines are still
+/// coming in faster than they can actually be flushed, when we reach this limit
+/// we'll pause and let it finish flushing rather than letting the queue
+/// continue growing.
+const SYNC_FLUSH_QD_LINES_THRESHOLD: usize = 10 * ASYNC_FLUSH_QD_LINES_THRESHOLD;
+
+/// Logging thread flushes at least this often.
+const MIN_FLUSH_FREQUENCY: Duration = Duration::from_secs(10);
+
+static SHADOW_LOGGER: Lazy<ShadowLogger> = Lazy::new(|| ShadowLogger::new());
+
+/// Helper for formatting times.
+struct TimeParts {
+    hours: u32,
+    mins: u32,
+    secs: u64,
+    nanos: u64,
+}
+
+impl TimeParts {
+    fn from_nanos(total_nanos: u128) -> Self {
+        let total_secs = u64::try_from(total_nanos / 1_000_000_000).unwrap();
+        let total_mins = u32::try_from(total_secs / 60).unwrap();
+        let hours = total_mins / 60;
+        let mins = total_mins - hours * 60;
+        let secs = total_secs - u64::from(total_mins) * 60;
+        let nanos = u64::try_from(total_nanos - u128::from(total_secs) * 1_000_000_000).unwrap();
+        Self {
+            hours,
+            mins,
+            secs,
+            nanos,
+        }
+    }
+}
+
+/// Initialize the Shadow logger.
+pub fn init() -> Result<(), SetLoggerError> {
+    log::set_logger(&*SHADOW_LOGGER)?;
+
+    // Start the thread that will receive log records and flush them to output.
+    std::thread::Builder::new()
+        .name("shadow-logger".to_string())
+        .spawn(move || SHADOW_LOGGER.logger_thread_fn())
+        .unwrap();
+
+    // Arrange to flush the logger on panic.
+    let default_panic_handler = std::panic::take_hook();
+    std::panic::set_hook(Box::new(move |panic_info| {
+        // Attempt to flush the logger. We want to avoid a recursive panic, so
+        // we flush the queue on the current thread instead of trying to send to
+        // send a command to the logger thread (because our thread-local sender
+        // may have already been destructed, and because the logger thread
+        // itself may be in a bad state), and ignore errors.
+        SHADOW_LOGGER.flush_records(None).ok();
+        default_panic_handler(panic_info);
+    }));
+
+    Ok(())
+}
+
+/// A logger specialized for Shadow. It attaches simulation context to log
+/// entries (e.g. sim time, running process, etc.). It's also designed for
+/// high performance to accomodate heavy logging from multiple threads.
+pub struct ShadowLogger {
+    // Channel used to send commands to second commands to the logger's thread.
```

Comment:
`send commands to second commands`: it's unclear what this means.

Reply:
Done